File: golang/tests/pkg/agentbay/agentbay_test.go
package agentbay_test

import (
	"fmt"
	"os"
	"testing"

	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
	"github.com/aliyun/wuying-agentbay-sdk/golang/tests/pkg/agentbay/testutil"
)

func TestNewAgentBay(t *testing.T) {
	// Test with API key provided directly
	testAPIKey := testutil.GetTestAPIKey(t)
	client, err := agentbay.NewAgentBay(testAPIKey)
	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}
	if client.APIKey != testAPIKey {
		t.Errorf("Expected API key to be '%s', got '%s'", testAPIKey, client.APIKey)
	}

	// Test with API key from environment variable
	os.Setenv("AGENTBAY_API_KEY", "env_api_key")
	defer os.Unsetenv("AGENTBAY_API_KEY")

	client, err = agentbay.NewAgentBay("")
	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}
	if client.APIKey != "env_api_key" {
		t.Errorf("Expected API key to be 'env_api_key', got '%s'", client.APIKey)
	}

	// Test with no API key
	os.Unsetenv("AGENTBAY_API_KEY")
	_, err = agentbay.NewAgentBay("")
	if err == nil {
		t.Fatal("Expected error for missing API key, got nil")
	}
}

func TestAgentBay_Create_List_Delete(t *testing.T) {
	// Initialize AgentBay client
	apiKey := testutil.GetTestAPIKey(t)
	agentBay, err := agentbay.NewAgentBay(apiKey)
	if err != nil {
		t.Fatalf("Error initializing AgentBay client: %v", err)
	}

	// Create a session
	fmt.Println("Creating a new session...")
	result, err := agentBay.Create(nil)
	if err != nil {
		t.Fatalf("Error creating session: %v", err)
	}

	// Check if RequestID exists
	if result.RequestID == "" {
		t.Logf("Warning: Expected non-empty RequestID")
	} else {
		t.Logf("Request ID: %s", result.RequestID)
	}

	session := result.Session
	t.Logf("Session created with ID: %s", session.SessionID)

	// Ensure session ID is not empty
	if session.SessionID == "" {
		t.Errorf("Expected non-empty session ID")
	}

	// List all sessions
	fmt.Println("Listing sessions...")
	listResult, err := agentBay.ListByLabels(nil)
	if err != nil {
		t.Fatalf("Error listing sessions: %v", err)
	}

	sessionIds := listResult.SessionIds

	// Ensure at least one session (the one we just created)
	if len(sessionIds) < 1 {
		t.Errorf("Expected at least 1 session, got %d", len(sessionIds))
	}

	// Check if our created session is in the list
	found := false
	for _, sessionId := range sessionIds {
		if sessionId == session.SessionID {
			found = true
			break
		}
	}
	if !found {
		t.Errorf("Created session with ID %s not found in sessions list", session.SessionID)
	}

	// Delete the session
	fmt.Println("Deleting the session...")
	deleteResult, err := agentBay.Delete(session)
	if err != nil {
		t.Fatalf("Error deleting session: %v", err)
	}

	// Check delete result
	if !deleteResult.Success {
		t.Errorf("Delete operation reported as unsuccessful")
	}

	// List sessions again to ensure it's deleted
	listResult, err = agentBay.ListByLabels(nil)
	if err != nil {
		t.Fatalf("Error listing sessions after deletion: %v", err)
	}

	sessionIds = listResult.SessionIds

	// Check if the deleted session is not in the list
	for _, sessionId := range sessionIds {
		if sessionId == session.SessionID {
			t.Errorf("Session with ID %s still exists after deletion", session.SessionID)
		}
	}
}

// TestAgentBay_ListByLabels tests the functionality of listing sessions by labels
func TestAgentBay_ListByLabels(t *testing.T) {
	// Initialize AgentBay client
	apiKey := testutil.GetTestAPIKey(t)
	agentBayClient, err := agentbay.NewAgentBay(apiKey)
	if err != nil {
		t.Fatalf("Error initializing AgentBay client: %v", err)
	}

	// Define two sets of labels for different sessions
	labelsA := map[string]string{
		"environment": "development",
		"owner":       "team-a",
		"project":     "project-x",
	}

	labelsB := map[string]string{
		"environment": "testing",
		"owner":       "team-b",
		"project":     "project-y",
	}

	// Create session with labels A
	t.Log("Creating session with labels A...")
	paramsA := agentbay.NewCreateSessionParams().WithLabels(labelsA)
	resultA, err := agentBayClient.Create(paramsA)
	if err != nil {
		t.Logf("Error creating session with labels A: %v", err)
		t.Skip("Skipping test as session creation failed")
		return
	}

	sessionA := resultA.Session
	t.Logf("Session A created with ID: %s", sessionA.SessionID)

	// Ensure cleanup of session A
	defer func() {
		t.Log("Cleaning up session A...")
		_, err := agentBayClient.Delete(sessionA)
		if err != nil {
			t.Logf("Warning: Error deleting session A: %v", err)
		}
	}()

	// Create session with labels B
	t.Log("Creating session with labels B...")
	paramsB := agentbay.NewCreateSessionParams().WithLabels(labelsB)
	resultB, err := agentBayClient.Create(paramsB)
	if err != nil {
		t.Logf("Error creating session with labels B: %v", err)
		t.Skip("Skipping test as session creation failed")
		return
	}

	sessionB := resultB.Session
	t.Logf("Session B created with ID: %s", sessionB.SessionID)

	// Ensure cleanup of session B
	defer func() {
		t.Log("Cleaning up session B...")
		_, err := agentBayClient.Delete(sessionB)
		if err != nil {
			t.Logf("Warning: Error deleting session B: %v", err)
		}
	}()

	// Test 1: List all sessions
	t.Log("Listing all sessions...")
	listResult, err := agentBayClient.ListByLabels(nil)
	if err != nil {
		t.Fatalf("Error listing all sessions: %v", err)
	}

	sessionIds := listResult.SessionIds
	t.Logf("Found %d sessions in total", len(sessionIds))

	// Test 2: List sessions by environment=development label
	t.Log("Listing sessions with environment=development...")
	devParams := agentbay.NewListSessionParams()
	devParams.Labels = map[string]string{"environment": "development"}
	devResult, err := agentBayClient.ListByLabels(devParams)
	if err != nil {
		t.Logf("Error listing sessions by environment=development: %v", err)
	} else {
		devSessionIds := devResult.SessionIds
		t.Logf("Found %d sessions with environment=development", len(devSessionIds))

		// Verify that session A is in the results
		foundSessionA := false
		for _, sessionId := range devSessionIds {
			if sessionId == sessionA.SessionID {
				foundSessionA = true
				break
			}
		}

		if !foundSessionA {
			t.Errorf("Expected to find session A in development environment results")
		}
	}

	// Test 3: List sessions by owner=team-b label
	t.Log("Listing sessions with owner=team-b...")
	teamBParams := agentbay.NewListSessionParams()
	teamBParams.Labels = map[string]string{"owner": "team-b"}
	teamBResult, err := agentBayClient.ListByLabels(teamBParams)
	if err != nil {
		t.Logf("Error listing sessions by owner=team-b: %v", err)
	} else {
		teamBSessionIds := teamBResult.SessionIds
		t.Logf("Found %d sessions with owner=team-b", len(teamBSessionIds))

		// Verify that session B is in the results
		foundSessionB := false
		for _, sessionId := range teamBSessionIds {
			if sessionId == sessionB.SessionID {
				foundSessionB = true
				break
			}
		}

		if !foundSessionB {
			t.Errorf("Expected to find session B in team-b owner results")
		}
	}

	// Test 4: List sessions with multiple labels (environment=testing AND project=project-y)
	t.Log("Listing sessions with environment=testing AND project=project-y...")
	multiParams := agentbay.NewListSessionParams()
	multiParams.Labels = map[string]string{
		"environment": "testing",
		"project":     "project-y",
	}
	multiResult, err := agentBayClient.ListByLabels(multiParams)
	if err != nil {
		t.Logf("Error listing sessions by multiple labels: %v", err)
	} else {
		multiLabelSessionIds := multiResult.SessionIds
		t.Logf("Found %d sessions with environment=testing AND project=project-y", len(multiLabelSessionIds))

		// Verify that session B is in the results and session A is not
		foundSessionA := false
		foundSessionB := false
		for _, sessionId := range multiLabelSessionIds {
			if sessionId == sessionA.SessionID {
				foundSessionA = true
			}
			if sessionId == sessionB.SessionID {
				foundSessionB = true
			}
		}

		if foundSessionA {
			t.Errorf("Did not expect to find session A in multi-label results")
		}
		if !foundSessionB {
			t.Errorf("Expected to find session B in multi-label results")
		}
	}

	// Test 5: List sessions with non-existent label
	t.Log("Listing sessions with non-existent label...")
	nonExistentParams := agentbay.NewListSessionParams()
	nonExistentParams.Labels = map[string]string{"non-existent": "value"}
	nonExistentResult, err := agentBayClient.ListByLabels(nonExistentParams)
	if err != nil {
		t.Logf("Error listing sessions by non-existent label: %v", err)
	} else {
		nonExistentSessionIds := nonExistentResult.SessionIds
		t.Logf("Found %d sessions with non-existent label", len(nonExistentSessionIds))
		if len(nonExistentSessionIds) > 0 {
			t.Logf("Warning: Found sessions with non-existent label, this might indicate an issue")
		}
	}

	t.Log("Test completed successfully")
}

// TestAgentBay_CreateSessionWithRecyclePolicy tests creating a session with custom RecyclePolicy
func TestAgentBay_CreateSessionWithRecyclePolicy(t *testing.T) {
	// Initialize AgentBay client
	apiKey := testutil.GetTestAPIKey(t)
	agentBay, err := agentbay.NewAgentBay(apiKey)
	if err != nil {
		t.Fatalf("Error initializing AgentBay client: %v", err)
	}

	fmt.Println("Testing session creation with custom RecyclePolicy...")

	// Create custom RecyclePolicy with Lifecycle1Day and default paths
	recyclePolicy := &agentbay.RecyclePolicy{
		Lifecycle: agentbay.Lifecycle1Day,
		Paths:     []string{""}, // Default path (applies to all paths)
	}

	// Validate the RecyclePolicy
	if err := recyclePolicy.Validate(); err != nil {
		t.Fatalf("RecyclePolicy validation failed: %v", err)
	}

	// Create custom SyncPolicy with the RecyclePolicy
	syncPolicy := &agentbay.SyncPolicy{
		UploadPolicy:   agentbay.NewUploadPolicy(),
		DownloadPolicy: agentbay.NewDownloadPolicy(),
		DeletePolicy:   agentbay.NewDeletePolicy(),
		ExtractPolicy:  agentbay.NewExtractPolicy(),
		RecyclePolicy:  recyclePolicy,
		BWList: &agentbay.BWList{
			WhiteLists: []*agentbay.WhiteList{
				{
					Path:         "",
					ExcludePaths: []string{},
				},
			},
		},
	}

	// Create ContextSync with the custom policy
	contextSync, err := agentbay.NewContextSync("test-recycle-context", "/tmp/recycle/path", syncPolicy)
	if err != nil {
		t.Fatalf("Error creating ContextSync: %v", err)
	}

	fmt.Printf("Created ContextSync with RecyclePolicy lifecycle: %s\n", recyclePolicy.Lifecycle)
	fmt.Printf("RecyclePolicy paths: %v\n", recyclePolicy.Paths)

	// Create session parameters with custom RecyclePolicy
	params := agentbay.NewCreateSessionParams().
		WithLabels(map[string]string{
			"test":      "recyclePolicy",
			"lifecycle": "1day",
		}).
		AddContextSyncConfig(contextSync)

	// Create session with custom RecyclePolicy
	fmt.Println("Creating session with custom RecyclePolicy...")
	result, err := agentBay.Create(params)
	if err != nil {
		t.Fatalf("Error creating session with RecyclePolicy: %v", err)
	}

	// Verify SessionResult structure
	if result.RequestID == "" {
		fmt.Println("Warning: Expected non-empty RequestID")
	} else {
		fmt.Printf("Create Session RequestID: %s\n", result.RequestID)
	}

	if result.Session == nil {
		t.Fatalf("Expected session to be created, got nil")
	}

	session := result.Session
	fmt.Printf("Session created successfully with ID: %s\n", session.SessionID)

	// Verify session properties
	if session.SessionID == "" {
		t.Errorf("Expected non-empty session ID")
	}

	// Verify that the session was created with the correct labels
	expectedLabels := map[string]string{
		"test":      "recyclePolicy",
		"lifecycle": "1day",
	}

	// List sessions with the labels to verify they were applied
	listParams := agentbay.NewListSessionParams()
	listParams.Labels = expectedLabels
	listResult, err := agentBay.ListByLabels(listParams)
	if err != nil {
		fmt.Printf("Warning: Error listing sessions by labels: %v\n", err)
	} else {
		found := false
		for _, sessionId := range listResult.SessionIds {
			if sessionId == session.SessionID {
				found = true
				break
			}
		}
		if !found {
			fmt.Println("Warning: Created session not found in filtered list")
		} else {
			fmt.Println("Session successfully found in filtered list with correct labels")
		}
	}

	// Ensure cleanup
	defer func() {
		fmt.Println("Cleaning up session with custom RecyclePolicy...")
		deleteResult, err := agentBay.Delete(session)
		if err != nil {
			fmt.Printf("Warning: Error deleting session: %v\n", err)
		} else {
			fmt.Printf("Delete Session RequestID: %s\n", deleteResult.RequestID)
		}
	}()

	fmt.Println("Session with custom RecyclePolicy created and verified successfully")
}
// TestRecyclePolicy_InvalidPaths tests invalid paths with wildcards via NewContextSync
func TestRecyclePolicy_InvalidPaths(t *testing.T) {
	fmt.Println("Testing invalid paths with wildcards...")

	// Use only one invalid path instead of for loop
	invalidPath := "/invalid/path/*"

	invalidRecyclePolicy := &agentbay.RecyclePolicy{
		Lifecycle: agentbay.Lifecycle1Day,
		Paths:     []string{invalidPath},
	}

	invalidSyncPolicy := &agentbay.SyncPolicy{
		UploadPolicy:   agentbay.NewUploadPolicy(),
		DownloadPolicy: agentbay.NewDownloadPolicy(),
		DeletePolicy:   agentbay.NewDeletePolicy(),
		ExtractPolicy:  agentbay.NewExtractPolicy(),
		RecyclePolicy:  invalidRecyclePolicy,
		BWList: &agentbay.BWList{
			WhiteLists: []*agentbay.WhiteList{
				{
					Path:         "",
					ExcludePaths: []string{},
				},
			},
		},
	}

	// This should fail at NewContextSync due to validation
	_, err := agentbay.NewContextSync("invalid-path-context", "/tmp/path", invalidSyncPolicy)
	if err == nil {
		t.Errorf("Expected invalid path '%s' to fail NewContextSync validation, but it passed", invalidPath)
	} else {
		fmt.Printf("Invalid path '%s' correctly failed NewContextSync validation: %v\n", invalidPath, err)
		
		// Verify error message contains expected information
		expectedSubstrings := []string{
			"wildcard patterns are not supported",
			invalidPath,
			"Please use exact directory paths instead",
		}
		
		for _, substring := range expectedSubstrings {
			if !contains(err.Error(), substring) {
				t.Errorf("Error message for invalid path '%s' should contain '%s', but got: %v", invalidPath, substring, err)
			}
		}
	}

	fmt.Println("Invalid paths with wildcards test completed successfully")
}

// TestRecyclePolicy_CombinedInvalid tests combination of invalid Lifecycle and invalid paths via NewContextSync
func TestRecyclePolicy_CombinedInvalid(t *testing.T) {
	fmt.Println("Testing combination of invalid Lifecycle and invalid paths...")

	combinedInvalidRecyclePolicy := &agentbay.RecyclePolicy{
		Lifecycle: "invalid_lifecycle",
		Paths:     []string{"/invalid/path/*"},
	}

	combinedInvalidSyncPolicy := &agentbay.SyncPolicy{
		UploadPolicy:   agentbay.NewUploadPolicy(),
		DownloadPolicy: agentbay.NewDownloadPolicy(),
		DeletePolicy:   agentbay.NewDeletePolicy(),
		ExtractPolicy:  agentbay.NewExtractPolicy(),
		RecyclePolicy:  combinedInvalidRecyclePolicy,
		BWList: &agentbay.BWList{
			WhiteLists: []*agentbay.WhiteList{
				{
					Path:         "",
					ExcludePaths: []string{},
				},
			},
		},
	}

	// This should fail at NewContextSync due to validation
	_, err := agentbay.NewContextSync("combined-invalid-context", "/tmp/path", combinedInvalidSyncPolicy)
	if err == nil {
		t.Errorf("Expected policy with both invalid lifecycle and invalid path to fail NewContextSync validation, but it passed")
	} else {
		fmt.Printf("Policy with both invalid lifecycle and invalid path correctly failed NewContextSync validation: %v\n", err)
		// Should fail on lifecycle validation first
		if !contains(err.Error(), "invalid lifecycle value") {
			t.Errorf("Expected error to mention invalid lifecycle value, but got: %v", err)
		}
	}

	fmt.Println("Combined invalid configuration test completed successfully")
}

// Helper function to check if a string contains a substring
func contains(s, substr string) bool {
	return len(s) >= len(substr) && (s == substr || len(substr) == 0 || 
		(len(s) > len(substr) && (s[:len(substr)] == substr || s[len(s)-len(substr):] == substr || 
		func() bool {
			for i := 0; i <= len(s)-len(substr); i++ {
				if s[i:i+len(substr)] == substr {
					return true
				}
			}
			return false
		}())))
}

File: typescript/tests/integration/agent-bay.test.ts
import {
  AgentBay,
  Session,
  AuthenticationError,
  APIError,
  ListSessionParams,
} from "../../src";
import { ContextSync, SyncPolicy, newSyncPolicy, Lifecycle } from "../../src/context-sync";
import { getTestApiKey } from "../utils/test-helpers";
import { log } from "../../src/utils/logger";

// Define Node.js process if it's not available
declare namespace NodeJS {
  interface ProcessEnv {
    [key: string]: string | undefined;
  }
}

declare var process: {
  env: {
    [key: string]: string | undefined;
  };
};

describe("AgentBay", () => {
  describe("constructor", () => {
    it("should initialize with API key from options", () => {
      const apiKey = getTestApiKey();
      const agentBay = new AgentBay({ apiKey });
      log(apiKey);

      expect(agentBay.getAPIKey()).toBe(apiKey);
    });

    it("should initialize with API key from environment variable", () => {
      const originalEnv = process.env.AGENTBAY_API_KEY;
      process.env.AGENTBAY_API_KEY = "env_api_key";

      try {
        const agentBay = new AgentBay();
        expect(agentBay.getAPIKey()).toBe("env_api_key");
      } finally {
        // Restore original environment
        if (originalEnv) {
          process.env.AGENTBAY_API_KEY = originalEnv;
        } else {
          delete process.env.AGENTBAY_API_KEY;
        }
      }
    });

    it("should throw AuthenticationError if no API key is provided", () => {
      const originalEnv = process.env.AGENTBAY_API_KEY;
      delete process.env.AGENTBAY_API_KEY;

      try {
        expect(() => new AgentBay()).toThrow(AuthenticationError);
      } finally {
        // Restore original environment
        if (originalEnv) {
          process.env.AGENTBAY_API_KEY = originalEnv;
        }
      }
    });
  });

  describe("create, list, and delete", () => {
    let agentBay: AgentBay;
    let session: Session;

    beforeEach(() => {
      const apiKey = getTestApiKey();
      agentBay = new AgentBay({ apiKey });
    });

    it("should create, list, and delete a session with requestId", async () => {
      // Create a session
      log("Creating a new session...");
      const createResponse = await agentBay.create();

      // Verify SessionResult structure
      expect(createResponse.success).toBe(true);
      expect(createResponse.requestId).toBeDefined();
      expect(typeof createResponse.requestId).toBe("string");
      expect(createResponse.requestId!.length).toBeGreaterThan(0);
      expect(createResponse.session).toBeDefined();
      expect(createResponse.errorMessage).toBeUndefined();

      session = createResponse.session!;
      log(`Session created with ID: ${session.sessionId}`);
      log(
        `Create Session RequestId: ${createResponse.requestId || "undefined"}`
      );

      // Ensure session ID is not empty
      expect(session.sessionId).toBeDefined();
      expect(session.sessionId.length).toBeGreaterThan(0);

      // List sessions
      log("Listing sessions...");
      const listResult = await agentBay.list();

      // Ensure at least one session (the one we just created)
      expect(listResult.sessionIds.length).toBeGreaterThanOrEqual(1);

      // Check if our created session is in the list
      const found = listResult.sessionIds.some((sid) => sid === session.sessionId);
      expect(found).toBe(true);

      // Delete the session
      log("Deleting the session...");
      const deleteResponse = await agentBay.delete(session);
      log(
        `Delete Session RequestId: ${deleteResponse.requestId || "undefined"}`
      );

      // Verify DeleteResult structure
      expect(deleteResponse.success).toBe(true);
      expect(deleteResponse.requestId).toBeDefined();
      expect(typeof deleteResponse.requestId).toBe("string");
      expect(deleteResponse.requestId!.length).toBeGreaterThan(0);
      expect(deleteResponse.errorMessage).toBeUndefined();

      // List sessions again to ensure it's deleted
      const listResultAfterDelete = await agentBay.list();

      // Check if the deleted session is not in the list
      const stillExists = listResultAfterDelete.sessionIds.some(
        (sid) => sid === session.sessionId
      );
      expect(stillExists).toBe(false);
    });
  });

  describe("listByLabels", () => {
    let agentBay: AgentBay;
    let sessionA: Session;
    let sessionB: Session;

    beforeEach(async () => {
      try {
        const apiKey = getTestApiKey();
        agentBay = new AgentBay({ apiKey });
        const labelsA = {
          environment: "development",
          owner: "team-a",
          project: "project-x",
        };

        const labelsB = {
          environment: "testing",
          owner: "team-b",
          project: "project-y",
        };

        // Create session with labels A
        log("Creating session with labels A...");
        const createResponseA = await agentBay.create({ labels: labelsA });

        // Verify SessionResult structure
        expect(createResponseA.success).toBe(true);
        expect(createResponseA.session).toBeDefined();

        sessionA = createResponseA.session!;
        log(`Session created with ID: ${sessionA.sessionId}`);
        log(
          `Create Session A RequestId: ${
            createResponseA.requestId || "undefined"
          }`
        );

        // Create session with labels B
        const createResponseB = await agentBay.create({ labels: labelsB });

        // Verify SessionResult structure
        expect(createResponseB.success).toBe(true);
        expect(createResponseB.session).toBeDefined();

        sessionB = createResponseB.session!;
        log(`Session created with ID: ${sessionB.sessionId}`);
        log(
          `Create Session B RequestId: ${
            createResponseB.requestId || "undefined"
          }`
        );
      } catch (error) {
        log(`Failed to constructor: ${error}`);
      }
    });

    afterEach(async () => {
      // Clean up sessions
      log("Cleaning up sessions...");
      if (sessionA) {
        try {
          const deleteResponseA = await agentBay.delete(sessionA);
          log(
            `Delete Session A RequestId: ${
              deleteResponseA.requestId || "undefined"
            }`
          );
        } catch (error) {
          log(`Warning: Error deleting session A: ${error}`);
        }
      }

      if (sessionB) {
        try {
          const deleteResponseB = await agentBay.delete(sessionB);
          log(
            `Delete Session B RequestId: ${
              deleteResponseB.requestId || "undefined"
            }`
          );
        } catch (error) {
          log(`Warning: Error deleting session B: ${error}`);
        }
      }
    });

    it("should list sessions by labels with requestId", async () => {
      // Test 1: List all sessions
      const allSessionsResult = await agentBay.list();
      log(`Found ${allSessionsResult.sessionIds.length} sessions in total`);

      // Test 2: List sessions by environment=development label using new API
      try {
        const devSessionsParams: ListSessionParams = {
          labels: { environment: "development" },
          maxResults: 5,
        };
        const devSessionsResponse = await agentBay.listByLabels(
          devSessionsParams
        );
        log(
          `List Sessions by environment=development RequestId: ${
            devSessionsResponse.requestId || "undefined"
          }`
        );
        log(
          `Total count: ${devSessionsResponse.totalCount}, Max results: ${devSessionsResponse.maxResults}`
        );

        // Verify SessionListResult structure
        expect(devSessionsResponse.success).toBe(true);
        expect(devSessionsResponse.requestId).toBeDefined();
        expect(typeof devSessionsResponse.requestId).toBe("string");
        expect(devSessionsResponse.requestId!.length).toBeGreaterThan(0);
        expect(devSessionsResponse.data).toBeDefined();
        expect(Array.isArray(devSessionsResponse.data)).toBe(true);

        // Verify that session A is in the results
        const foundSessionA = devSessionsResponse.data.some(
          (s) => s.sessionId === sessionA.sessionId
        );
        expect(foundSessionA).toBe(true);
      } catch (error) {
        log(`Error listing sessions by environment=development: ${error}`);
      }

      // Test 3: List sessions by owner=team-b label using new API
      try {
        const teamBSessionsParams: ListSessionParams = {
          labels: { owner: "team-b" },
          maxResults: 5,
        };
        const teamBSessionsResponse = await agentBay.listByLabels(
          teamBSessionsParams
        );
        log(
          `List Sessions by owner=team-b RequestId: ${
            teamBSessionsResponse.requestId || "undefined"
          }`
        );
        log(
          `Total count: ${teamBSessionsResponse.totalCount}, Max results: ${teamBSessionsResponse.maxResults}`
        );

        // Verify SessionListResult structure
        expect(teamBSessionsResponse.success).toBe(true);
        expect(teamBSessionsResponse.requestId).toBeDefined();

        // Verify that session B is in the results
        const foundSessionB = teamBSessionsResponse.data.some(
          (s) => s.sessionId === sessionB.sessionId
        );
        expect(foundSessionB).toBe(true);
      } catch (error) {
        log(`Error listing sessions by owner=team-b: ${error}`);
      }

      // Test 4: List sessions with multiple labels (environment=testing AND project=project-y) using new API
      try {
        const multiLabelSessionsParams: ListSessionParams = {
          labels: {
            environment: "testing",
            project: "project-y",
          },
          maxResults: 5,
        };
        const multiLabelSessionsResponse = await agentBay.listByLabels(
          multiLabelSessionsParams
        );
        log(
          `Found ${multiLabelSessionsResponse.data.length} sessions with environment=testing AND project=project-y`
        );
        log(
          `Total count: ${multiLabelSessionsResponse.totalCount}, Max results: ${multiLabelSessionsResponse.maxResults}`
        );
        log(
          `List Sessions by multiple labels RequestId: ${
            multiLabelSessionsResponse.requestId || "undefined"
          }`
        );

        // Verify SessionListResult structure
        expect(multiLabelSessionsResponse.success).toBe(true);
        expect(multiLabelSessionsResponse.requestId).toBeDefined();

        // Verify that session B is in the results and session A is not
        const foundSessionA = multiLabelSessionsResponse.data.some(
          (s) => s.sessionId === sessionA.sessionId
        );
        const foundSessionB = multiLabelSessionsResponse.data.some(
          (s) => s.sessionId === sessionB.sessionId
        );

        expect(foundSessionA).toBe(false);
        expect(foundSessionB).toBe(true);

        // Demonstrate pagination if there's a next token
        if (multiLabelSessionsResponse.nextToken) {
          log("\nFetching next page...");
          const nextPageParams: ListSessionParams = {
            ...multiLabelSessionsParams,
            nextToken: multiLabelSessionsResponse.nextToken,
          };
          const nextPageResponse = await agentBay.listByLabels(nextPageParams);
          log(`Next page sessions count: ${nextPageResponse.data.length}`);
          log(`Next page RequestId: ${nextPageResponse.requestId}`);

          // Verify next page result structure
          expect(nextPageResponse.success).toBe(true);
          expect(nextPageResponse.requestId).toBeDefined();
        }
      } catch (error) {
        log(`Error listing sessions by multiple labels: ${error}`);
      }

      // Test 5: List sessions with non-existent label using new API
      try {
        const nonExistentSessionsParams: ListSessionParams = {
          labels: { "non-existent": "value" },
          maxResults: 5,
        };
        const nonExistentSessionsResponse = await agentBay.listByLabels(
          nonExistentSessionsParams
        );
        log(
          `Found ${nonExistentSessionsResponse.data.length} sessions with non-existent label`
        );
        log(
          `Total count: ${nonExistentSessionsResponse.totalCount}, Max results: ${nonExistentSessionsResponse.maxResults}`
        );
        log(
          `List Sessions by non-existent label RequestId: ${
            nonExistentSessionsResponse.requestId || "undefined"
          }`
        );

        // Verify SessionListResult structure even for empty results
        expect(nonExistentSessionsResponse.success).toBe(true);
        expect(nonExistentSessionsResponse.requestId).toBeDefined();

        if (nonExistentSessionsResponse.data.length > 0) {
          log(
            "Warning: Found sessions with non-existent label, this might indicate an issue"
          );
        }
      } catch (error) {
        log(`Error listing sessions by non-existent label: ${error}`);
      }

      log("Test completed successfully");
    });
  });

  describe("create session with custom recyclePolicy", () => {
    let agentBay: AgentBay;
    let session: Session;

    beforeEach(() => {
      const apiKey = getTestApiKey();
      agentBay = new AgentBay({ apiKey });
    });

    afterEach(async () => {
      // Clean up session
      if (session) {
        try {
          log("Cleaning up session with custom recyclePolicy...");
          const deleteResponse = await agentBay.delete(session);
          log(
            `Delete Session RequestId: ${
              deleteResponse.requestId || "undefined"
            }`
          );
        } catch (error) {
          log(`Warning: Error deleting session: ${error}`);
        }
      }
    });

    it("should create session with custom recyclePolicy using Lifecycle_1Day", async () => {
      // Create custom recyclePolicy with Lifecycle_1Day and default paths
      const customSyncPolicy: SyncPolicy = {
        ...newSyncPolicy(),
        recyclePolicy: {
          lifecycle: Lifecycle.Lifecycle_1Day,
          paths: [""] // Using default path value
        }
      };

      // Create ContextSync with custom policy
      const contextSync = new ContextSync(
        "test-recycle-context",
        "/test/recycle/path",
        customSyncPolicy
      );

      log("Creating session with custom recyclePolicy...");
      log(`RecyclePolicy lifecycle: ${customSyncPolicy.recyclePolicy?.lifecycle}`);
      log(`RecyclePolicy paths: ${JSON.stringify(customSyncPolicy.recyclePolicy?.paths)}`);

      // Create session with custom recyclePolicy
      const createResponse = await agentBay.create({
        labels: { test: "recyclePolicy", lifecycle: "1day" },
        contextSync: [contextSync]
      });

      // Verify SessionResult structure
      expect(createResponse.success).toBe(true);
      expect(createResponse.requestId).toBeDefined();
      expect(typeof createResponse.requestId).toBe("string");
      expect(createResponse.requestId!.length).toBeGreaterThan(0);
      expect(createResponse.session).toBeDefined();
      expect(createResponse.errorMessage).toBeUndefined();

      session = createResponse.session!;
      log(`Session created successfully with ID: ${session.sessionId}`);
      log(`Create Session RequestId: ${createResponse.requestId || "undefined"}`);

      // Verify session properties
      expect(session.sessionId).toBeDefined();
      expect(session.sessionId.length).toBeGreaterThan(0);

      log("Session with custom recyclePolicy created and verified successfully");
    });

    it("should throw error when creating ContextSync with invalid recyclePolicy path", () => {
      // Create custom recyclePolicy with invalid wildcard path
      const invalidSyncPolicy: SyncPolicy = {
        ...newSyncPolicy(),
        recyclePolicy: {
          lifecycle: Lifecycle.Lifecycle_1Day,
          paths: ["/invalid/path/*"] // Invalid path with wildcard
        }
      };

      log("Testing ContextSync creation with invalid recyclePolicy path...");
      log(`Invalid path: ${invalidSyncPolicy.recyclePolicy?.paths[0]}`);

      // Test that ContextSync constructor throws an error for invalid recyclePolicy path
      expect(() => {
        new ContextSync(
          "test-invalid-context",
          "/test/path",
          invalidSyncPolicy
        );
      }).toThrow("Wildcard patterns are not supported in path. Got: /invalid/path/*. Please use exact directory paths instead.");

      log("ContextSync correctly threw error for invalid recyclePolicy path");
    });

    it("should throw error when creating ContextSync with invalid lifecycle", () => {
      log("Testing ContextSync creation with invalid lifecycle...");

      // Create custom recyclePolicy with invalid lifecycle value
      const invalidSyncPolicy: SyncPolicy = {
        ...newSyncPolicy(),
        recyclePolicy: {
          lifecycle: "invalid_lifecycle" as any, // Invalid lifecycle
          paths: [""]
        }
      };

      log(`Invalid lifecycle: ${invalidSyncPolicy.recyclePolicy?.lifecycle}`);

      // Test that ContextSync constructor throws an error for invalid lifecycle
      expect(() => {
        new ContextSync(
          "invalid-lifecycle-context",
          "/test/path",
          invalidSyncPolicy
        );
      }).toThrow(/Invalid lifecycle value: invalid_lifecycle\. Valid values are:/);

      log("ContextSync correctly threw error for invalid lifecycle");
    });

    it("should throw error when creating ContextSync with combined invalid configuration", () => {
      log("Testing ContextSync creation with combined invalid lifecycle and invalid paths...");

      // Create custom recyclePolicy with both invalid lifecycle and invalid path
      const combinedInvalidSyncPolicy: SyncPolicy = {
        ...newSyncPolicy(),
        recyclePolicy: {
          lifecycle: "invalid_lifecycle" as any, // Invalid lifecycle
          paths: ["/invalid/path/*"] // Invalid path with wildcard
        }
      };

      log(`Invalid lifecycle: ${combinedInvalidSyncPolicy.recyclePolicy?.lifecycle}`);
      log(`Invalid path: ${combinedInvalidSyncPolicy.recyclePolicy?.paths[0]}`);

      // Test that ContextSync constructor throws an error (should fail on lifecycle validation first)
      expect(() => {
        new ContextSync(
          "combined-invalid-context",
          "/test/path",
          combinedInvalidSyncPolicy
        );
      }).toThrow(/Invalid lifecycle value: invalid_lifecycle\. Valid values are:/);

      log("ContextSync correctly threw error for combined invalid configuration");
    });
  });
});